library/std/src/sys/windows/args.rs

```diff
@@ -234,3 +236,152 @@ impl Iterator for WStrUnits<'_> {
         }
     }
 }
+
+#[derive(Debug)]
+pub(crate) enum Arg {
+    /// Add quotes (if needed)
+    Regular(OsString),
+    /// Append raw string without quoting
+    Raw(OsString),
+}
+
+enum Quote {
+    // Every arg is quoted
+    Always,
+    // Whitespace and empty args are quoted
+    Auto,
+    // Arg appended without any changes (#29494)
+    Never,
+}
+
+pub(crate) fn append_arg(cmd: &mut Vec<u16>, arg: &Arg, force_quotes: bool) -> io::Result<()> {
+    let (arg, quote) = match arg {
+        Arg::Regular(arg) => (arg, if force_quotes { Quote::Always } else { Quote::Auto }),
+        Arg::Raw(arg) => (arg, Quote::Never),
+    };
+
+    // If an argument has 0 characters then we need to quote it to ensure
+    // that it actually gets passed through on the command line or otherwise
+    // it will be dropped entirely when parsed on the other end.
+    ensure_no_nuls(arg)?;
+    let arg_bytes = arg.bytes();
+    let (quote, escape) = match quote {
+        Quote::Always => (true, true),
+        Quote::Auto => {
+            (arg_bytes.iter().any(|c| *c == b' ' || *c == b'\t') || arg_bytes.is_empty(), true)
+        }
+        Quote::Never => (false, false),
+    };
+    if quote {
+        cmd.push('"' as u16);
+    }
+
+    let mut backslashes: usize = 0;
+    for x in arg.encode_wide() {
+        if escape {
+            if x == '\\' as u16 {
+                backslashes += 1;
+            } else {
+                if x == '"' as u16 {
+                    // Add n+1 backslashes to total 2n+1 before internal '"'.
+                    cmd.extend((0..=backslashes).map(|_| '\\' as u16));
+                }
+                backslashes = 0;
+            }
+        }
+        cmd.push(x);
+    }
+
+    if quote {
+        // Add n backslashes to total 2n before ending '"'.
+        cmd.extend((0..backslashes).map(|_| '\\' as u16));
+        cmd.push('"' as u16);
+    }
+    Ok(())
+}
+
+pub(crate) fn make_bat_command_line(
+    script: &[u16],
+    args: &[Arg],
+    force_quotes: bool,
+) -> io::Result<Vec<u16>> {
+    // Set the start of the command line to `cmd.exe /c "`
+    // It is necessary to surround the command in an extra pair of quotes,
+    // hence The trailing quote here. It will be closed after all arguments
+    // have been added.
+    let mut cmd: Vec<u16> = "cmd.exe /c \"".encode_utf16().collect();
+
+    // Push the script name surrounded by its quote pair.
+    cmd.push(b'"' as u16);
+    cmd.extend_from_slice(script.strip_suffix(&[0]).unwrap_or(script));
```

Comment:
Could this use `append_arg`? What happens if the script name needs quoting?

Reply:
It will only need quoting if it contains a `"`, which cannot occur in Windows file names. The problem with `append_arg` is it requires a `OsString` whereas at this point we have the UTF-16 string that was used to test if the file exists. I admit I'm somewhat reluctant to keep switching between encodings.

For now maybe an alternative is to return an error if the script name does in fact contain `"`?

Reply:
Can `"` occur in filenames that get interpreted by a driver (e.g. a filesystem driver for a foreign filesystem)?

Could a `\` ever occur at the *end* of a filename that could be executable? Is it theoretically possible to achieve a filename ending in `\` using reparse points, or NTFS Streams, or special `\\` paths? If so, would the `\` escape the `"`?

Reply:
In [Naming Files, Paths, and Namespaces](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file) search for "reserved characters". There's a list of characters that are reserved for special use by the OS. These cannot be used in filenames at all. Indeed, `argv[0]` depends upon this fact as a [`"` cannot be escaped in `argv[0]`](https://github.com/rust-lang/rust/blob/4ce257ff198d23bdf14e956fbf2fe0fed297201f/library/std/src/sys/windows/args.rs#L79).

A trailing `\` is guarded against [earlier](https://github.com/rust-lang/rust/blob/4ce257ff198d23bdf14e956fbf2fe0fed297201f/library/std/src/sys/windows/process.rs#L375) and I cannot think of a way that such a filename could be possible to returned from the OS. Though I guess double checking wouldn't hurt.

Reply:
I adjusted the last commit to do just that.